File: 150-evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.py
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack = []
        ops = {"+", "-", "*", "/"}

        for token in tokens:
            if token in ops:
                b = stack.pop()
                a = stack.pop()

                if token == "+":
                    stack.append(a + b)
                elif token == "-":
                    stack.append(a - b)
                elif token == "*":
                    stack.append(a * b)
                elif token == "/":
                    stack.append(int(a / b))
            else:
                stack.append(int(token))
        
        return stack.pop()

File: 150-evaluate-reverse-polish-notation/evaluate-reverse-polish-notation.java
class Solution {
    public int evalRPN(String[] tokens) {
    
    Deque<Integer> stack = new ArrayDeque<>();
    // using ArrayDeque under the hood to implement Stack
    // faster because its unsynchronized

    Set<String> ops = Set.of("+", "-", "*", "/");
    // Set of valid operators

    for(String token : tokens){

        if(ops.contains(token)){
            int b = stack.pop();
            int a = stack.pop();
            int result = 0;

            switch(token){
                case "+": result = a + b; break;
                case "-": result = a - b; break;
                case "*": result = a * b; break;
                case "/": result = a / b; break; 
                // integer division truncates towards zero by default
            }
            stack.push(result);
        } else {
            stack.push(Integer.parseInt(token));
        }
    }
    return stack.pop();
    // final result
  }
}